wandb/sdk/lib/apikey.py

```
@@ -170,7 +182,33 @@ def prompt_api_key(  # noqa: C901
     return key
 
 
-def write_netrc(host: str, entity: str, key: str) -> Optional[bool]:
+def check_netrc_access(
+    netrc_path: str,
+) -> _NetrcPermissions:
+    """Check if we can read and write to the netrc file."""
+    file_exists = False
+    write_access = False
+    read_access = False
+    try:
+        st = os.stat(netrc_path)
+        file_exists = True
+        write_access = bool(st.st_mode & stat.S_IWUSR)
+        read_access = bool(st.st_mode & stat.S_IRUSR)
+    except FileNotFoundError:
+        # If the netrc file doesn't exist, we will create it.
+        write_access = True
+        read_access = True
+    except OSError as e:
+        wandb.termerror(f"Unable to read permissions for {netrc_path}, {e}")
+
+    return _NetrcPermissions(
+        exists=file_exists,
+        write_access=write_access,
+        read_access=read_access,
+    )
+
+
+def write_netrc(host: str, entity: str, key: str) -> bool:
```

The `bool` return value is not actually used except in a test. What do you think of letting errors bubble out and handling them higher up?

The call hierarchy is:
* apikey.py: write_netrc
* apikey.py: write_key
* wandb_login.py: configure_api_key
* wandb_login.py: _login

Does it make sense for `write_key` to silently return without writing a key? I think probably not. The call in `_login` looks like this:

```py
    if not key_is_pre_configured:
        wlogin.configure_api_key(key)
        wlogin.update_session(key, status=key_status)
        wlogin._update_global_anonymous_setting()
```

(The naming is off here: What does "configure_api_key" do? It should be split into "warn_if_in_notebook" and "write_key") `update_session` and so on should still run even if writing a key fails, so we could either handle the error in `_login` or rename `configure_api_key` ("try_save_key"?) and handle the error there

---

I think I like the idea of using `try_save_key` and raising an error from write_netrc.
Since we do not want to completely fail login, when a user still might be executing other commands in the same process.